src/app/ReadClient.cpp

```diff
@@ -253,7 +266,51 @@ CHIP_ERROR ReadClient::SendReadRequest(ReadPrepareParams & aReadPrepareParams)
             }
         }
 
-        ReturnErrorOnFailure(request.IsFabricFiltered(aReadPrepareParams.mIsFabricFiltered).EndOfReadRequestMessage().GetError());
+        ReturnErrorOnFailure(request.IsFabricFiltered(aReadPrepareParams.mIsFabricFiltered).GetError());
+
+        bool hasEncodeDataVersionList = false;
+        if (!attributePaths.empty())
+        {
+            uint32_t number = 0;
+            TLV::TLVWriter backup;
+            request.Checkpoint(backup);
+            if (!dataVersionFilters.empty())
+            {
+                DataVersionFilterIBs::Builder & dataVersionFilterListBuilder = request.CreateDataVersionFilters();
+                ReturnErrorOnFailure(request.GetError());
+                ReturnErrorOnFailure(
+                    GenerateDataVersionFilterList(dataVersionFilterListBuilder, attributePaths, dataVersionFilters, number));
+                if (number == 0)
+                {
+                    request.Rollback(backup);
+                }
+                else
+                {
+                    hasEncodeDataVersionList = true;
+                }
+            }
+            else
+            {
+                DataVersionFilterIBs::Builder & dataVersionFilterListBuilder = request.CreateDataVersionFilters();
```

Comment:
If we do this in both branches of the if/else, why not just put it before the `if`?

Reply:
good catch, thanks, updated.